Module generation for FNAL UPS package management  system

marcmengel opened the pull request:
We add two module types: `ups_table` and `ups_version` that should be enabled together to get ups declarations in `$SPACK_ROOT/share/spack/upsdb` by default.

marcmengel committed:
trying to hook it all up

marcmengel committed:
trying to hook it all up

marcmengel committed:
trying to hook it all up

marcmengel committed:
trying to hook it all up

marcmengel committed:
trying to hook it all up

marcmengel committed:
trying to hook it all up

marcmengel committed:
pretty close to right, need to generate flavors...

marcmengel committed:
pretty close to right, need to generate flavors...

marcmengel committed:
pretty close to right, need to generate flavors...

marcmengel committed:
pretty close to right, need to generate flavors...

marcmengel committed:
pretty close to right, need to generate flavors...

marcmengel committed:
pretty close to right, need to generate flavors...

marcmengel committed:
pretty close to right, need to generate flavors...

marcmengel committed:
pretty close to right, need to generate flavors...

marcmengel committed:
now with flavor and qualifier conversions and basic dependencies, still need same on dependencies

marcmengel committed:
now with flavor and qualifier conversions and basic dependencies, still need same on dependencies

marcmengel committed:
now with flavor and qualifier conversions and basic dependencies, still need same on dependencies

marcmengel committed:
now with flavor and qualifier conversions and basic dependencies, still need same on dependencies

marcmengel committed:
now with flavor and qualifier conversions and basic dependencies, still need same on dependencies

marcmengel committed:
now with flavor and qualifier conversions and basic dependencies, still need same on dependencies

marcmengel committed:
now with flavor and qualifier conversions and basic dependencies, still need same on dependencies

marcmengel committed:
now with flavor and qualifier conversions and basic dependencies, still need same on dependencies

marcmengel committed:
one more bit

marcmengel committed:
one more bit

marcmengel committed:
one more bit

marcmengel committed:
one more bit

marcmengel committed:
one more bit

marcmengel committed:
one more bit

marcmengel committed:
Now with versions on dependencies

marcmengel committed:
Now with versions on dependencies

marcmengel committed:
Now with versions on dependencies

marcmengel committed:
Now with versions on dependencies

marcmengel committed:
don't forget the actula module files

marcmengel committed:
don't forget the actula module files

marcmengel committed:
don't forget the actula module files

marcmengel committed:
don't forget the actula module files

marcmengel committed:
remove is okay after all(?)

marcmengel committed:
remove is okay after all(?)

marcmengel committed:
remove is okay after all(?)

marcmengel committed:
remove is okay after all(?)

marcmengel committed:
use UPS_PROD_DIR where applicable

marcmengel committed:
use UPS_PROD_DIR where applicable

marcmengel committed:
use UPS_PROD_DIR where applicable

marcmengel committed:
use UPS_PROD_DIR where applicable

marcmengel committed:
Merge branch 'ups_gen' of ssh://cdcvs.fnal.gov/cvs/projects/spack-infrastructure-spack into ups_gen

marcmengel committed:
Merge branch 'ups_gen' of ssh://cdcvs.fnal.gov/cvs/projects/spack-infrastructure-spack into ups_gen

marcmengel committed:
per-instance table files, and single qualifiers in table files

marcmengel committed:
per-instance table files, and single qualifiers in table files

marcmengel committed:
debug cleanup

marcmengel committed:
debug cleanup

marcmengel committed:
now works on MVP1a

marcmengel committed:
now works on MVP1a

greenc-FNAL commented:
1. In order to have any chance of getting this merged upstream, it's going to need tests.
1. Likewise: does it pass `spack flake8`?
1. I'm a little uneasy that one has to enable _two_ module types in order to get the full set of desired behavior. Is this due to a limitation in the core code? Could you write a focused pitch on what would need to be done in core to allow one module type to generate two files?

marcmengel commented:
I'm not sure we want this accepted upstream... It's a local patch for the
migration period.
It is also fairly baked-in that the module types generate a single file per
module type. for example:
-- the interface has a "filename(self)" method to extract the (singular)
filename of the generated file; and
-- the template for the module type is named modulefile.template_type
One would have to make the code do something like check the result type of
filename() and accept either the current filename string, or perhaps a map
of {"template_filename": "/where/to/put/it", "template_filename2":
"/where/to/put/it"} to support multiple file generation; or else modify all
of the existing module types to return such a map.
Of course, not having it accepted upstream is slightly painful in its own
way...

On Fri, Aug 2, 2019 at 10:11 AM Chris Green <notifications@github.com>
wrote:

>
>    1. In order to have any chance of getting this merged upstream, it's
>    going to need tests.
>    2. Likewise: does it pass spack flake8?
>    3. I'm a little uneasy that one has to enable *two* module types in
>    order to get the full set of desired behavior. Is this due to a limitation
>    in the core code? Could you write a focused pitch on what would need to be
>    done in core to allow one module type to generate two files?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/FNALssi/spack/pull/21?email_source=notifications&email_token=AAB6GV33RZ2LBZ3SXFD42M3QCRFBLA5CNFSM4II5LE32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3OAYLI#issuecomment-517737517>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAB6GV4Y5P7N3BW4LILZO2TQCRFBLANCNFSM4II5LE3Q>
> .
>

greenc-FNAL commented:
The migration period could last literal years, and I really wouldn't look forward to maintaining this as part of a fork for long given observed activity levels and the fact that this affects quite a few core files. It's been bad enough dealing with relatively localized changes that have taken a few months to get into upstream core.

Who calls `filename()` and for what purpose is the information used?

greenc-FNAL commented on lib/spack/llnl/util/tty/__init__.py:
```diff
@@ -169,7 +169,8 @@ def info(message, *args, **kwargs):
     break_long_words = kwargs.get('break_long_words', False)
-    st_countback = kwargs.get('countback', 3)
+    #st_countback = kwargs.get('countback', 3)
+    st_countback = 10
```
Was this intended for debug that should have been removed?

greenc-FNAL commented on lib/spack/spack/cmd/module.py:
```diff
@@ -10,6 +10,8 @@
 import spack.cmd.modules.dotkit
 import spack.cmd.modules.lmod
 import spack.cmd.modules.tcl
+import spack.cmd.modules.ups_version
```
Once https://github.com/spack/spack/pull/11145 gets merged upstream I was planning on submitting a PR to use it here, removing the need for the top level command to know about every sub-command. See https://github.com/FNALssi/spack-dev/blob/master/dev/cmd/dev.py for an idea of how this might work.

marcmengel committed:
tests half roughed in

marcmengel committed:
tests half roughed in

marcmengel committed:
fun with tests

marcmengel committed:
fun with tests

marcmengel committed:
test fun

marcmengel committed:
test fun

marcmengel committed:
Merge branch 'ups_gen' of github.com:marcmengel/spack into ups_gen

marcmengel commented:
I've added tests for the ups_version and ups_table module types, and rebased off upstream's develop.
If you bring fnal_develop up to upstream's develop it should be a clean merge now.

gartung commented:
@marcmengel If you change the base of the pull request to spack/spack/develop it would be easier to see just your changes.

greenc-FNAL commented:
@gartung: @marcmengel has changed the base branch to make the review easier.

greenc-FNAL commented:
@marcmengel this should now merge clean against `fnal-develop`, so I changed the base to that. Would you be able to update the title and description to something that would provide a meaningful commit message for the squashed merge? Alternatively if you'd like to manually squash commits into a small number of thematically coherent commits, we could do a straight merge. Please let me know.

greenc-FNAL commented:
@marcmengel Ping on the [`filename()` question](https://github.com/FNALssi/spack/pull/21#issuecomment-517754960).

marcmengel commented:
> @marcmengel Ping on the [`filename()` question](https://github.com/FNALssi/spack/pull/21#issuecomment-517754960).

So filename() is called to get the name of the module file that we're generating.  I specialize it here to make it put the table file and/or version file in the right place relative to the ups database directory, which is specified as the root for both module types in the config.
